tools/syz-declextract/run.go

```diff
@@ -122,6 +151,237 @@ type context struct {
+		fmt.Printf("%v: %v is not mapped to any file\n", name, fops)
+		return
+	}
+	fmt.Printf("%v: %v mapped to %v\n", name, fops, files[:min(10, len(files))])
+
+	// Some fops are mapped to too many files, usually these have generic callbacks
+	// like simple_attr_read+simple_attr_write. Compiler restricts number of strings to 2000.
+	// TODO: emit multiple open calls to cover all files.
+	files = files[:min(len(files), 1000)]
+	w := new(bytes.Buffer)
+	fmt.Fprintf(w, "\n# %v\n", fops)
+	fmt.Fprintf(w, "resource fd_%v[fd]\n", name)
+	fileFlags := fmt.Sprintf("\"%s\"", files[0])
+	if len(files) > 1 {
+		fileFlags = fmt.Sprintf("%v_files", name)
+		fmt.Fprintf(w, "%v = ", fileFlags)
+		for i, file := range files {
+			if i != 0 {
+				fmt.Fprintf(w, ", ")
+			}
+			fmt.Fprintf(w, "\"%v\"", file)
+		}
+		fmt.Fprintf(w, "\n")
+	}
+	fmt.Fprintf(w, "openat$%v(fd const[AT_FDCWD], file ptr[in, string[%v]],"+
+		" flags flags[open_flags], mode const[0]) fd_%v (automatic)\n",
+		name, fileFlags, name)
+	if fops.Read != "" {
+		fmt.Fprintf(w, "read$%v(fd fd_%v, buf ptr[out, array[int8]],"+
+			" len bytesize[buf]) (automatic)\n", name, name)
+	}
+	if fops.Write != "" {
+		fmt.Fprintf(w, "write$%v(fd fd_%v, buf ptr[in, array[int8]],"+
+			" len bytesize[buf]) (automatic)\n", name, name)
+	}
+	if fops.Mmap != "" {
+		fmt.Fprintf(w, "mmap$%v(addr vma, len len[addr], prot flags[mmap_prot],"+
+			" flags flags[mmap_flags], fd fd_%v, offset fileoff) (automatic)\n", name, name)
+	}
+	if fops.Ioctl != "" {
+		if len(fops.Cmds) == 0 {
+			fmt.Fprintf(w, "ioctl$%v(fd fd_%v, cmd intptr,"+
+				" arg ptr[in, array[int8]]) (automatic)\n", name, name)
+		} else {
+			for _, cmd := range fops.Cmds {
+				suffix := ""
+				ctx.ioctlNames[cmd.Name]++
+				if ctx.ioctlNames[cmd.Name] != 1 {
+					suffix += fmt.Sprint(ctx.ioctlNames[cmd.Name])
+				}
+				fmt.Fprintf(w, "ioctl$auto_%v%v(fd fd_%v, cmd const[%v], arg %v) (automatic)\n",
+					cmd.Name, suffix, name, cmd.Name, cmd.Type)
+			}
+		}
+	}
+	fmt.Fprintf(w, "\n")
+
+	parsed := ast.Parse(w.Bytes(), "", nil)
+	if parsed == nil {
+		panic(fmt.Sprintf("parsing failed:\n%s", w.Bytes()))
+	}
+	ctx.nodes = append(ctx.nodes, parsed.Nodes...)
+	ctx.nodes = append(ctx.nodes, &ast.NewLine{})
+}
+
+func (ctx *context) mapFopsToFiles(fops *OutputFops) (string, []string) {
+	first := true
+	var files map[string]bool
+	var unique map[string]int
+	for _, fn := range []string{fops.Open, fops.Read, fops.Write, fops.Mmap, fops.Ioctl} {
+		if fn == "" {
+			continue
+		}
+		files1 := ctx.probeFuncToFiles[fn]
+		var unique1 map[string]int
+		if fn != "seq_read" && !strings.HasPrefix(fn, "generic_") &&
+			!strings.HasPrefix(fn, "simple_") {
+			unique1 = make(map[string]int)
+			for i, part := range strings.Split(fn, "_") {
+				switch part {
+				case "read", "write", "ioctl", "mmap", "open", "fops":
+					continue
+				}
+				unique1[part] = i
+			}
+			if unique == nil {
+				unique = unique1
+			} else {
+				for part := range unique {
```

Comment:
Here you essentially select the part(s) of the first considered function name (most likely that in `fops.Open`) that are not encountered in any other function name, right? 

In this case I don't quite understand how it worked e.g. here:
```
# fops: open:crc_control_open read:seq_read write:crc_control_write
resource fd_auto_crc_control[fd]
```

and here
```
fops: open:tracing_trace_options_open read:seq_read write:tracing_trace_options_write
resource fd_auto_tracing_trace_options[fd]
```

Overall, it feels like this approach relies on lots and lots of names being just very indiscriptive -- otherwise it would have made sense to look for the common parts of the names, not for the unique ones.

I wonder if the file names could tell us more? At least the dirname of them?